[wasm] Add backward compatibility file to timezone data compilation

tqiu8 opened the pull request:
backward file in runtime-assets
Fix #43518

tqiu8 committed:
Add backward file from iana to handle backward compatibility in tz data

tqiu8 committed:
Add backward file from iana to handle backward compatibility in tz data

tqiu8 committed:
Merge branch 'master' into local-tz

CoffeeFlux commented:
Shouldn't `backward` also have been added [here](https://github.com/dotnet/runtime-assets/blob/master/src/System.Runtime.TimeZoneData/System.Runtime.TimeZoneData.csproj)?

tqiu8 commented:
It's included on this [line](https://github.com/dotnet/runtime-assets/blob/master/src/System.Runtime.TimeZoneData/System.Runtime.TimeZoneData.csproj#L26). Could be included in the Area ItemGroup but the separation was just for clarity

lewing commented:
likely blocked on https://github.com/dotnet/runtime/pull/43768

lewing reviewed (commented):
Change looks good, let's get some tests in place as well.

CoffeeFlux commented:
Ah, missed that, thanks.

This is also a case where I'd really like a test, though I'm not sure how hard that is with the Browser.

tqiu8 committed:
add tests to check for backwards compatibility

tqiu8 committed:
add tests to check for backwards compatibility

tqiu8 committed:
add tests to check for backwards compatibility

tqiu8 committed:
add tests to check for backwards compatibility

tqiu8 committed:
add platform specific to tests

tqiu8 committed:
add platform specific to tests

tqiu8 committed:
add clear cached data at the end of test to prevent affecting other tests

tqiu8 committed:
add clear cached data at the end of test to prevent affecting other tests

lewing reviewed (commented):
We need to wait for the runtime-assets update in https://github.com/dotnet/runtime/pull/43768 but the changes look good assuming
the tests pass.  Might be worth it to add checks for the actual tz offset for completeness but these checks would catch the failure case and offset checks are potentially more fragile (they could however validate the assets changes)

radical commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -2321,6 +2321,46 @@ public static void TestNameWithInvariantCulture()
+            Assert.Equal(tz.StandardName, localtz.StandardName);
+            Assert.Equal(tz.DisplayName, localtz.DisplayName); 
+
+            TimeZoneInfo.ClearCachedData();
+            Environment.SetEnvironmentVariable("TZ", "");
```
This should be in a `finally{}` block, so that we ensure that it is always cleared, and TZ is reset.

lewing replied:
Strictly speaking it should probably save the original TZ value and restore it in the finally block.  We run the risk of breaking some other test otherwise right?

dotnet-maestro[bot] committed:
[master] Update dependencies from dotnet/arcade Microsoft/vstest dotnet/llvm-project dotnet/runtime-assets mono/linker (#43768)

[master] Update dependencies from dotnet/arcade Microsoft/vstest dotnet/llvm-project dotnet/runtime-assets mono/linker

radical commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -2321,6 +2321,46 @@ public static void TestNameWithInvariantCulture()
         }
 
+        [Theory]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
```
Is this test specific to wasm? IIUC, the `backwards` file is being added only for that? Then maybe we should use `[PlatformSpecific(TestPlatforms.Browser)]`?

lewing replied:
This isn't really specific to browser although it would be unlikely to fail anywhere else.

radical commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -2321,6 +2321,46 @@ public static void TestNameWithInvariantCulture()
+        [InlineData("America/Jujuy")]
+        [InlineData("America/Mendoza")]
+        [InlineData("America/Indianapolis")]
+        public static void ChangeLocalTimeZone(string id) 
```
This test is checking that setting the `TZ` environment var leads to the correct tz being returned, is that correct? Unrelated to this PR?

Maybe rename the test to `ReadLocalTZFromEnvironmentVariable`

lewing replied:
Wasm currently uses uses the TZ setting via shared code with unix and this tests the exact case that was failing for browser so it is very related.

tqiu8 replied:
Changing the `TZ` environment variable changes the local timezone. We're checking that when we change the local timezone with different ids that the correct one is still returned.

tqiu8 committed:
PR feedback

tqiu8 committed:
PR feedback

tqiu8 committed:
Merge branch 'local-tz' of github.com:tqiu8/runtime into local-tz

tqiu8 committed:
set TZ var to original value instead of empty

tqiu8 committed:
move original tz out of try

lewing reviewed (approved):
Looks good.

am11 commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -2321,6 +2321,50 @@ public static void TestNameWithInvariantCulture()
+        }
+
+        [Theory]
+        [PlatformSpecific(TestPlatforms.Browser)]
```
afaik, the general practice is to not restrict test cases for particular platform, when they pass on other platforms.

tqiu8 commented:
/backport to release/5.0
